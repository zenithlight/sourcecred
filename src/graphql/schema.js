// @flow

/**
 * Data types to describe a particular subset of GraphQL schemata.
 * Schemata represented by this module must satisfy these constraints:
 *
 *   - Every object must have an `id` field of primitive type.
 *   - Every field of an object must be either a primitive, a reference
 *     to a single (possibly nullable) object, or a _connection_ as
 *     described in the Relay cursor connections specification. In
 *     particular, no field may directly contain a list.
 *   - Interface types must be represented as unions of all their
 *     implementations.
 */

// The name of a GraphQL type, like `Repository` or `Int`.
export type Typename = string;

// The name of a GraphQL object field, like `name` or `pullRequests`.
export type Fieldname = string;

// The database-wide unique ID of a GraphQL object.
export type ObjectId = string;

// Description of a GraphQL schema. Types are represented as follows:
//   - An object type is represented directly as an `OBJECT`.
//   - A union type is represented directly as a `UNION`.
//   - An interface type is represented as a `UNION` of all its
//     implementations.
//   - Scalars and enums may only occur as object fields, and are
//     represented as `PRIMITIVE`s (except for `ID`s).
//   - Connections are supported as object fields, but arbitrary lists
//     are not.
//
// Primitive and enum fields on an object type may optionally be
// annotated with their representations.
//
// To accommodate schemata where some object types do not have IDs,
// objects may have "nested" fields of primitive or node-reference type.
// These may be nested to depth exactly 1. Suppose that `Foo` is an
// object type that includes `bar: Bar!`, but `Bar` is an object type
// without an `id`. Then `Bar` may not be a first-class type, but `Foo`
// may pull properties off of it using
//
//     bar: nested({x: primitive(), y: node("Baz")});
//
// The property "bar" in the above example is called a _nested_
// property, and its fields "x" and "y" are called _eggs_. (The nest
// contains the eggs.)
export type Schema = {+[Typename]: NodeType};
export type NodeType =
  | {|+type: "SCALAR", +representation: "string" | "number" | "boolean"|}
  | {|+type: "ENUM", +values: {|+[string]: true|}|}
  | {|+type: "OBJECT", +fields: {|+[Fieldname]: FieldType|}|}
  | {|+type: "UNION", +clauses: {|+[Typename]: true|}|};

export type FieldType =
  | IdFieldType
  | PrimitiveFieldType
  | NodeFieldType
  | ConnectionFieldType
  | NestedFieldType;
export type IdFieldType = {|+type: "ID"|};
export type PrimitiveFieldType = {|
  +type: "PRIMITIVE",
  +annotation: null | PrimitiveTypeAnnotation,
|};
export type PrimitiveTypeAnnotation = {|
  +nonNull: boolean,
  +elementType: Typename,
|};
export type NodeFieldType = {|
  +type: "NODE",
  +elementType: Typename,
  +fidelity: Fidelity,
|};
export type ConnectionFieldType = {|
  +type: "CONNECTION",
  +elementType: Typename,
  +fidelity: Fidelity,
|};
export type NestedFieldType = {|
  +type: "NESTED",
  +eggs: {+[Fieldname]: PrimitiveFieldType | NodeFieldType},
|};

// A field is _faithful_ if selecting its `__typename` and `id` will
// always yield the correct `__typename` for the node of the given ID.
// In theory, this should always be the case, but some remote schemas
// are broken. For details, see:
//
//   - https://github.com/sourcecred/sourcecred/issues/996
//   - https://github.com/sourcecred/sourcecred/issues/998
//
// For an unfaithful field, the `actualTypenames` property lists all the
// types of objects that can _actually_ be returned when the field is
// queried. (This set only affects generated Flow types, not runtime
// semantics.) These must all be object types.
//
// It is always sound to represent an actually-faithful field as
// unfaithful, but doing so may incur additional queries. Marking a type
// as faithful should be seen as an optimization that may be performed
// only when the server is abiding by its contract for that field.
export type Fidelity =
  | {|+type: "FAITHFUL"|}
  | {|+type: "UNFAITHFUL", actualTypenames: {|+[Typename]: true|}|};

export function faithful(): Fidelity {
  return {type: "FAITHFUL"};
}

export function unfaithful(
  actualTypenames: $ReadOnlyArray<Typename>
): Fidelity {
  const actualTypenamesObject: {|[Typename]: true|} = ({}: any);
  for (const t of actualTypenames) {
    actualTypenamesObject[t] = true;
  }
  return {type: "UNFAITHFUL", actualTypenames: actualTypenamesObject};
}

// Every object must have exactly one `id` field, and it must have this
// name.
const ID_FIELD_NAME = "id";

export function schema(types: {[Typename]: NodeType}): Schema {
  function assertKind(
    path,
    elementTypename,
    validKinds,
    {isFidelity = false} = {}
  ) {
    const self =
      (isFidelity ? "unfaithful typenames list of " : "") +
      `field ${path.map((x) => JSON.stringify(x)).join("/")}`;
    const elementType = types[elementTypename];
    if (elementType == null) {
      throw new Error(`${self} has unknown type: "${elementTypename}"`);
    }
    if (!validKinds.includes(elementType.type)) {
      throw new Error(
        `${self} has invalid type "${elementTypename}" ` +
          `of kind "${elementType.type}"`
      );
    }
  }
  function validateFidelity(path, fidelity) {
    switch (fidelity.type) {
      case "FAITHFUL":
        break;
      case "UNFAITHFUL":
        for (const typename of Object.keys(fidelity.actualTypenames)) {
          assertKind(path, typename, ["OBJECT"], {isFidelity: true});
        }
        break;
      // istanbul ignore next: unreachable per Flow
      default:
        throw new Error((fidelity.type: empty));
    }
  }
  const result = {};
  for (const typename of Object.keys(types)) {
    const type = types[typename];
    switch (type.type) {
      case "SCALAR":
        result[typename] = {
          type: "SCALAR",
          representation: type.representation,
        };
        break;
      case "ENUM":
        result[typename] = {type: "ENUM", values: {...type.values}};
        break;
      case "OBJECT":
        for (const fieldname of Object.keys(type.fields)) {
          const field = type.fields[fieldname];
          switch (field.type) {
            case "ID":
              // Nothing to check.
              break;
            case "PRIMITIVE":
              if (field.annotation != null) {
                assertKind(
                  [typename, fieldname],
                  field.annotation.elementType,
                  ["SCALAR", "ENUM"]
                );
              }
              break;
            case "NODE":
              assertKind([typename, fieldname], field.elementType, [
                "OBJECT",
                "UNION",
              ]);
              validateFidelity([typename, fieldname], field.fidelity);
              break;
            case "CONNECTION":
              assertKind([typename, fieldname], field.elementType, [
                "OBJECT",
                "UNION",
              ]);
              validateFidelity([typename, fieldname], field.fidelity);
              break;
            case "NESTED":
              for (const eggName of Object.keys(field.eggs)) {
                const egg = field.eggs[eggName];
                switch (egg.type) {
                  case "PRIMITIVE":
                    if (egg.annotation != null) {
                      assertKind(
                        [typename, fieldname, eggName],
                        egg.annotation.elementType,
                        ["SCALAR", "ENUM"]
                      );
                    }
                    break;
                  case "NODE":
                    assertKind(
                      [typename, fieldname, eggName],
                      egg.elementType,
                      ["OBJECT", "UNION"]
                    );
                    validateFidelity(
                      [typename, fieldname, eggName],
                      egg.fidelity
                    );
                    break;
                  // istanbul ignore next: unreachable per Flow
                  default:
                    throw new Error((egg.type: empty));
                }
              }
              break;
            // istanbul ignore next: unreachable per Flow
            default:
              throw new Error((field.type: empty));
          }
        }
        result[typename] = {type: "OBJECT", fields: {...type.fields}};
        break;
      case "UNION":
        for (const clause of Object.keys(type.clauses)) {
          const clauseType = types[clause];
          if (clauseType == null) {
            throw new Error(
              `union has unknown clause: "${typename}"/"${clause}"`
            );
          }
          if (clauseType.type !== "OBJECT") {
            // The GraphQL spec doesn't permit unions of interfaces or
            // other unions (or primitives). This is nice, because it
            // means that we don't have to worry about ill-founded
            // unions.
            throw new Error(
              `union has non-object type clause: "${typename}"/"${clause}"`
            );
          }
        }
        result[typename] = {type: "UNION", clauses: {...type.clauses}};
        break;
      // istanbul ignore next
      default:
        throw new Error((type.type: empty));
    }
  }
  return result;
}

export function scalar(
  representation: "string" | "number" | "boolean"
): NodeType {
  return {type: "SCALAR", representation};
}

function enum_(values: $ReadOnlyArray<string>): NodeType {
  const valuesObject: {|[string]: true|} = ({}: any);
  for (const v of values) {
    valuesObject[v] = true;
  }
  return {type: "ENUM", values: valuesObject};
}
export {enum_ as enum};

export function object(fields: {[Fieldname]: FieldType}): NodeType {
  for (const fieldname of Object.keys(fields)) {
    const field = fields[fieldname];
    if (fieldname === "__typename") {
      throw new Error("reserved field name: " + fieldname);
    }
    if (field.type === "ID" && fieldname !== ID_FIELD_NAME) {
      throw new Error(`invalid ID field with name "${fieldname}"`);
    }
  }
  if (fields[ID_FIELD_NAME] == null) {
    throw new Error(`expected ID field with name "${ID_FIELD_NAME}"`);
  }
  if (fields[ID_FIELD_NAME].type !== "ID") {
    throw new Error(`field "${ID_FIELD_NAME}" must be an ID field`);
  }
  // Workaround for <https://github.com/facebook/flow/issues/7128>.
  const exactFields: {|[Fieldname]: FieldType|} = ({...fields}: any);
  return {type: "OBJECT", fields: exactFields};
}

export function union(clauses: $ReadOnlyArray<Typename>): NodeType {
  const clausesMap: {|[Typename]: true|} = ({}: any);
  for (const clause of clauses) {
    if (clausesMap[clause] != null) {
      throw new Error(`duplicate union clause: "${clause}"`);
    }
    clausesMap[clause] = true;
  }
  return {type: "UNION", clauses: clausesMap};
}

export function id(): IdFieldType {
  return {type: "ID"};
}

export function primitive(
  annotation?: PrimitiveTypeAnnotation
): PrimitiveFieldType {
  return {type: "PRIMITIVE", annotation: annotation || null};
}

export function node(
  elementType: Typename,
  fidelity: Fidelity = faithful()
): NodeFieldType {
  return {type: "NODE", elementType, fidelity};
}

export function connection(
  elementType: Typename,
  fidelity: Fidelity = faithful()
): ConnectionFieldType {
  return {type: "CONNECTION", elementType, fidelity};
}

export function nonNull(elementType: Typename): PrimitiveTypeAnnotation {
  return {nonNull: true, elementType};
}

export function nullable(elementType: Typename): PrimitiveTypeAnnotation {
  return {nonNull: false, elementType};
}

export function nested(eggs: {
  +[Fieldname]: PrimitiveFieldType | NodeFieldType,
}): NestedFieldType {
  return {type: "NESTED", eggs: {...eggs}};
}
