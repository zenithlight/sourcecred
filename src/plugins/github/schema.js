// @flow

import * as Schema from "../../graphql/schema";

export default function schema(): Schema.Schema {
  const s = Schema;
  const types = {
    DateTime: s.scalar("string"),
    GitObjectID: s.scalar("string"),
    GitTimestamp: s.scalar("string"),
    Int: s.scalar("number"),
    String: s.scalar("string"),
    URI: s.scalar("string"),
    PullRequestReviewState: s.enum([
      "PENDING",
      "COMMENTED",
      "APPROVED",
      "CHANGES_REQUESTED",
      "DISMISSED",
    ]),
    ReactionContent: s.enum([
      "THUMBS_UP",
      "THUMBS_DOWN",
      "LAUGH",
      "HOORAY",
      "CONFUSED",
      "HEART",
      "ROCKET",
      "EYES",
    ]),
    Repository: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      name: s.primitive(s.nonNull("String")),
      owner: s.node("RepositoryOwner"),
      issues: s.connection("Issue"),
      pullRequests: s.connection("PullRequest"),
      defaultBranchRef: s.node("Ref"),
      createdAt: s.primitive(s.nonNull("DateTime")),
    }),
    Issue: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      title: s.primitive(s.nonNull("String")),
      body: s.primitive(s.nonNull("String")),
      number: s.primitive(s.nonNull("Int")),
      author: s.node("Actor"),
      comments: s.connection("IssueComment"),
      reactions: s.connection("Reaction"),
      createdAt: s.primitive(s.nonNull("DateTime")),
    }),
    PullRequest: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      title: s.primitive(s.nonNull("String")),
      body: s.primitive(s.nonNull("String")),
      number: s.primitive(s.nonNull("Int")),
      mergeCommit: s.node("Commit"),
      additions: s.primitive(s.nonNull("Int")),
      deletions: s.primitive(s.nonNull("Int")),
      author: s.node("Actor"),
      comments: s.connection("IssueComment"), // yes, PRs have IssueComments
      reviews: s.connection("PullRequestReview"),
      reactions: s.connection("Reaction"),
      createdAt: s.primitive(s.nonNull("DateTime")),
      baseRefName: s.primitive(s.nonNull("String")),
    }),
    IssueComment: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      body: s.primitive(s.nonNull("String")),
      author: s.node("Actor"),
      reactions: s.connection("Reaction"),
      createdAt: s.primitive(s.nonNull("DateTime")),
    }),
    PullRequestReview: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      body: s.primitive(s.nonNull("String")),
      author: s.node("Actor"),
      state: s.primitive(s.nonNull("PullRequestReviewState")),
      comments: s.connection("PullRequestReviewComment"),
      createdAt: s.primitive(s.nonNull("DateTime")),
    }),
    PullRequestReviewComment: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      body: s.primitive(s.nonNull("String")),
      author: s.node("Actor"),
      reactions: s.connection("Reaction"),
      createdAt: s.primitive(s.nonNull("DateTime")),
    }),
    Reaction: s.object({
      id: s.id(),
      content: s.primitive(s.nonNull("ReactionContent")),
      user: s.node("User", s.unfaithful(["User", "Organization"])),
      createdAt: s.primitive(s.nonNull("DateTime")),
    }),
    Ref: s.object({
      id: s.id(),
      // Unlike most node references, this is guaranteed non-null (but
      // we have no way to express that).
      target: s.node("GitObject"),
    }),
    GitObject: s.union(["Blob", "Commit", "Tag", "Tree"]),
    Blob: s.object({
      id: s.id(),
      oid: s.primitive(s.nonNull("GitObjectID")),
    }),
    Commit: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      oid: s.primitive(s.nonNull("GitObjectID")),
      message: s.primitive(s.nonNull("String")),
      author: /* GitActor */ s.nested({
        // The GitHub schema indicates that `date` can be null, but does
        // not indicate when this might be the case.
        date: s.primitive(s.nullable("GitTimestamp")),
        user: s.node("User", s.unfaithful(["User", "Bot"])),
      }),
      parents: s.connection("Commit"),
      // In contrast to the author.date, this is both nonNull and is
      // specifically the authoredDate. Docs for author.date suggest that
      // field might be the commiter date instead.
      authoredDate: s.primitive(s.nonNull("GitTimestamp")),
    }),
    Tag: s.object({
      id: s.id(),
      oid: s.primitive(s.nonNull("GitObjectID")),
    }),
    Tree: s.object({
      id: s.id(),
      oid: s.primitive(s.nonNull("GitObjectID")),
    }),
    Actor: s.union(["User", "Bot", "Organization"]), // actually an interface
    RepositoryOwner: s.union(["User", "Organization"]), // actually an interface
    User: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      login: s.primitive(s.nonNull("String")),
    }),
    Bot: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      login: s.primitive(s.nonNull("String")),
    }),
    Organization: s.object({
      id: s.id(),
      url: s.primitive(s.nonNull("URI")),
      login: s.primitive(s.nonNull("String")),
    }),
  };
  return s.schema(types);
}
