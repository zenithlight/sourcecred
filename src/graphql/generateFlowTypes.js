// @flow

import prettier, {type Options as PrettierOptions} from "prettier";

import type {
  ConnectionFieldType,
  Fidelity,
  FieldType,
  IdFieldType,
  NestedFieldType,
  NodeFieldType,
  PrimitiveFieldType,
  Schema,
  Typename,
} from "./schema";

export default function generateFlowTypes(
  schema: Schema,
  prettierOptions: PrettierOptions
): string {
  const definitions = [];

  function formatField(field: FieldType): string {
    switch (field.type) {
      case "ID":
        return formatIdField(field);
      case "PRIMITIVE":
        return formatPrimitiveField(field);
      case "NODE":
        return formatNodeField(field);
      case "CONNECTION":
        return formatConnectionField(field);
      case "NESTED":
        return formatNestedField(field);
      // istanbul ignore next: unreachable per Flow
      default:
        throw new Error((field.type: empty));
    }
  }
  function formatIdField(_unused_field: IdFieldType): string {
    return "string";
  }
  function formatFidelity(nominal: Typename, fidelity: Fidelity): string {
    switch (fidelity.type) {
      case "FAITHFUL":
        return nominal;
      case "UNFAITHFUL": {
        const values = Object.keys(fidelity.actualTypenames).sort();
        return values.length === 0 ? "empty" : values.join(" | ");
      }
      // istanbul ignore next: unreachable per Flow
      default:
        throw new Error((fidelity.type: empty));
    }
  }
  function formatPrimitiveField(field: PrimitiveFieldType): string {
    if (field.annotation == null) {
      return "mixed";
    } else if (field.annotation.nonNull) {
      return field.annotation.elementType;
    } else {
      return "null | " + field.annotation.elementType;
    }
  }
  function formatNodeField(field: NodeFieldType): string {
    const elementType = formatFidelity(field.elementType, field.fidelity);
    return `null | ${elementType}`;
  }
  function formatConnectionField(field: ConnectionFieldType): string {
    const elementType = formatFidelity(field.elementType, field.fidelity);
    return `$ReadOnlyArray<null | ${elementType}>`;
  }
  function formatNestedField(field: NestedFieldType): string {
    const eggs = [];
    for (const eggName of Object.keys(field.eggs).sort()) {
      eggs.push({eggName, rhs: formatField(field.eggs[eggName])});
    }
    const eggContents = eggs.map((x) => `+${x.eggName}: ${x.rhs}`).join(", ");
    return `null | {|\n${eggContents}\n|}`;
  }

  for (const typename of Object.keys(schema).sort()) {
    const type = schema[typename];
    switch (type.type) {
      case "SCALAR":
        definitions.push(`export type ${typename} = ${type.representation};`);
        break;
      case "ENUM": {
        const values = Object.keys(type.values)
          .sort()
          .map((x) => JSON.stringify(x));

        // export type E = "A" | "B";
        const typeRhs = values.length === 0 ? "empty" : values.join(" | ");
        definitions.push(`export type ${typename} = ${typeRhs};`);

        // export const E$Values: {|+A: "A", +B: "B"|} = deepFreeze(...);
        const objectName = `${typename}$Values`;
        const objectType = [
          "{|",
          values.map((x) => `+${x}: ${x}`).join(", "),
          "|}",
        ].join("");
        const objectValue = [
          "deepFreeze({",
          values.map((x) => `${x}: ${x}`).join(", "),
          "})",
        ].join("");
        definitions.push(
          `export const ${objectName}: ${objectType} = ${objectValue};`
        );
        break;
      }
      case "OBJECT": {
        const fields = [
          {fieldname: "__typename", rhs: JSON.stringify(typename)},
        ];
        for (const fieldname of Object.keys(type.fields).sort()) {
          fields.push({fieldname, rhs: formatField(type.fields[fieldname])});
        }
        const fieldContents = fields
          .map((x) => `+${x.fieldname}: ${x.rhs}`)
          .join(", ");
        const rhs = `{|\n${fieldContents}\n|}`;
        definitions.push(`export type ${typename} = ${rhs};`);
        break;
      }
      case "UNION": {
        const rhs =
          Object.keys(type.clauses).length === 0
            ? "empty"
            : Object.keys(type.clauses).sort().join(" | ");
        definitions.push(`export type ${typename} = ${rhs};`);
        break;
      }
      // istanbul ignore next: unreachable per Flow
      default:
        throw new Error((type.type: empty));
    }
  }

  const rawSource = [
    "// @flow",
    "// Autogenerated file. Do not edit.",
    'import deepFreeze from "deep-freeze"',
    ...definitions,
  ].join("\n\n");
  return prettier.format(rawSource, prettierOptions);
}
