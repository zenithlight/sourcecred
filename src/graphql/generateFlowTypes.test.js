// @flow

import generateFlowTypes from "./generateFlowTypes";
import * as Schema from "./schema";

describe("graphql/generateFlowTypes", () => {
  function run(schema: Schema.Schema): string {
    return generateFlowTypes(schema, {parser: "babel", trailingComma: "es5"});
  }

  describe("generateFlowTypes", () => {
    it("works on a representative schema", () => {
      // This schema should be constructed as to singlehandedly achieve
      // full test coverage.
      const s = Schema;
      const schema = s.schema({
        String: s.scalar("string"),
        DateTime: s.scalar("string"),
        Dollars: s.scalar("number"),
        Color: s.enum(["RED", "GREEN", "BLUE"]),
        PaintJob: s.object({
          id: s.id(),
          metadata: s.primitive(/* unannotated */),
          cost: s.primitive(s.nonNull("Dollars")),
          completed: s.primitive(s.nullable("DateTime")),
          color: s.primitive(s.nonNull("Color")),
          designer: s.node("Actor"),
          painters: s.connection("Actor"),
          client: s.node(
            "Human",
            s.unfaithful(["Human", "TalentedChimpanzee"])
          ),
          coClients: s.connection(
            "Human",
            s.unfaithful(["Human", "TalentedChimpanzee"])
          ),
          referrer: s.node("PaintJob"),
          relatedWork: s.connection("PaintJob"),
          details: s.nested({
            moreMetadata: s.primitive(),
            description: s.primitive(s.nullable("String")),
            oldColor: s.primitive(s.nonNull("Color")),
            oldPainter: s.node("Actor"),
            oldClient: s.node(
              "Human",
              s.unfaithful(["Human", "TalentedChimpanzee"])
            ),
          }),
        }),
        Actor: s.union(["Human", "TalentedChimpanzee"]),
        Human: s.object({
          id: s.id(),
          name: s.primitive(s.nullable("String")),
        }),
        TalentedChimpanzee: s.object({
          id: s.id(),
          name: s.primitive(s.nullable("String")),
        }),
        EmptyEnum: s.enum([]),
        EmptyUnion: s.union([]),
        EmptyUnfaithfulSet: s.object({
          id: s.id(),
          foo: s.node("Human", s.unfaithful([])),
        }),
      });
      expect(run(schema)).toMatchSnapshot();
    });

    it("respects the Prettier options", () => {
      const s = Schema;
      const schema = s.schema({
        E: s.enum(["ONE", "TWO"]),
      });
      const options1 = {parser: "babel", singleQuote: false};
      const options2 = {parser: "babel", singleQuote: true};
      const output1 = generateFlowTypes(schema, options1);
      const output2 = generateFlowTypes(schema, options2);
      expect(output1).not.toEqual(output2);
    });

    it("is invariant with respect to type definition order", () => {
      const s = Schema;
      const s1 = s.schema({
        A: s.object({id: s.id()}),
        B: s.object({id: s.id()}),
      });
      const s2 = s.schema({
        B: s.object({id: s.id()}),
        A: s.object({id: s.id()}),
      });
      expect(run(s1)).toEqual(run(s2));
    });

    it("is invariant with respect to enum clause order", () => {
      const s = Schema;
      const s1 = s.schema({
        E: s.enum(["ONE", "TWO"]),
      });
      const s2 = s.schema({
        E: s.enum(["TWO", "ONE"]),
      });
      expect(run(s1)).toEqual(run(s2));
    });

    it("is invariant with respect to object field order", () => {
      const s = Schema;
      const s1 = s.schema({
        O: s.object({
          id: s.id(),
          x: s.primitive(),
          y: s.primitive(),
        }),
      });
      const s2 = s.schema({
        O: s.object({
          id: s.id(),
          y: s.primitive(),
          x: s.primitive(),
        }),
      });
      expect(run(s1)).toEqual(run(s2));
    });

    it("is invariant with respect to nested egg order", () => {
      const s = Schema;
      const s1 = s.schema({
        O: s.object({
          id: s.id(),
          nest: s.nested({
            x: s.primitive(),
            y: s.primitive(),
          }),
        }),
      });
      const s2 = s.schema({
        O: s.object({
          id: s.id(),
          nest: s.nested({
            y: s.primitive(),
            x: s.primitive(),
          }),
        }),
      });
      expect(run(s1)).toEqual(run(s2));
    });

    it("is invariant with respect to unfaithful typename order", () => {
      const s = Schema;
      const s1 = s.schema({
        O: s.object({
          id: s.id(),
          x: s.node("A", s.unfaithful(["A", "B"])),
        }),
        A: s.object({id: s.id()}),
        B: s.object({id: s.id()}),
      });
      const s2 = s.schema({
        O: s.object({
          id: s.id(),
          x: s.node("A", s.unfaithful(["B", "A"])),
        }),
        A: s.object({id: s.id()}),
        B: s.object({id: s.id()}),
      });
      expect(run(s1)).toEqual(run(s2));
    });

    it("is invariant with respect to union clause order", () => {
      const s = Schema;
      const s1 = s.schema({
        A: s.object({id: s.id()}),
        B: s.object({id: s.id()}),
        U: s.union(["A", "B"]),
      });
      const s2 = s.schema({
        A: s.object({id: s.id()}),
        B: s.object({id: s.id()}),
        U: s.union(["B", "A"]),
      });
      expect(run(s1)).toEqual(run(s2));
    });
  });
});
