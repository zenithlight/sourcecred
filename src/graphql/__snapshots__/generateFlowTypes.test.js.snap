// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql/generateFlowTypes generateFlowTypes works on a representative schema 1`] = `
"// @flow

// Autogenerated file. Do not edit.

import deepFreeze from \\"deep-freeze\\";

export type Actor = Human | TalentedChimpanzee;

export type Color = \\"BLUE\\" | \\"GREEN\\" | \\"RED\\";

export const Color$Values: {|
  +BLUE: \\"BLUE\\",
  +GREEN: \\"GREEN\\",
  +RED: \\"RED\\",
|} = deepFreeze({ BLUE: \\"BLUE\\", GREEN: \\"GREEN\\", RED: \\"RED\\" });

export type DateTime = string;

export type Dollars = number;

export type EmptyEnum = empty;

export const EmptyEnum$Values: {||} = deepFreeze({});

export type EmptyUnfaithfulSet = {|
  +__typename: \\"EmptyUnfaithfulSet\\",
  +foo: null | empty,
  +id: string,
|};

export type EmptyUnion = empty;

export type Human = {|
  +__typename: \\"Human\\",
  +id: string,
  +name: null | String,
|};

export type PaintJob = {|
  +__typename: \\"PaintJob\\",
  +client: null | Human | TalentedChimpanzee,
  +coClients: $ReadOnlyArray<null | Human | TalentedChimpanzee>,
  +color: Color,
  +completed: null | DateTime,
  +cost: Dollars,
  +designer: null | Actor,
  +details: null | {|
    +description: null | String,
    +moreMetadata: mixed,
    +oldClient: null | Human | TalentedChimpanzee,
    +oldColor: Color,
    +oldPainter: null | Actor,
  |},
  +id: string,
  +metadata: mixed,
  +painters: $ReadOnlyArray<null | Actor>,
  +referrer: null | PaintJob,
  +relatedWork: $ReadOnlyArray<null | PaintJob>,
|};

export type String = string;

export type TalentedChimpanzee = {|
  +__typename: \\"TalentedChimpanzee\\",
  +id: string,
  +name: null | String,
|};
"
`;
